nni/compression/pytorch/quantization_speedup/integrated_tensorrt.py

```diff
@@ -44,346 +38,320 @@ def valid_config(config=None):
+    return weight_layer.name
+
+def apply_precision_to_layer(layer, config):
+    if 'weight_bits' in config:
+        w_bits = config['weight_bits']
         layer.precision = Precision_Dict[w_bits]
-        pre_layer.precision = Precision_Dict[w_bits]
-        next_layer.precision = Precision_Dict[w_bits]
-        # set the first two layers' output type
-        pre_layer.set_output_type(0, Precision_Dict[w_bits])
-        layer.set_output_type(0, Precision_Dict[w_bits])
-        pre_in_tensor = pre_layer.get_input(0)
+    if 'input_bits' in config:
+        assert 'tracked_min_input' in config
+        assert 'tracked_max_input' in config
+        tracked_min_input = config['tracked_min_input']
+        tracked_max_input = config['tracked_max_input']
         in_tensor = layer.get_input(0)
-        next_in_tensor = next_layer.get_input(0)
-        # set three layers' input tensor range
-        pre_in_tensor.dynamic_range = (tracked_min_input, tracked_max_input)
         in_tensor.dynamic_range = (tracked_min_input, tracked_max_input)
-        next_in_tensor.dynamic_range = (tracked_min_input, tracked_max_input)
-
-    # if output bits exists, set the last layer's output type output tensor range
-    if 'output_bits' in config[layer.name]:
-        assert 'tracked_min_output' in config[layer.name]
-        assert 'tracked_max_output' in config[layer.name]
-        a_bits = config[layer.name]['output_bits']
-        tracked_min_output = config[layer.name]['tracked_min_output']
-        tracked_max_output = config[layer.name]['tracked_max_output']
-        # set the last layer's output type
-        next_layer.set_output_type(0, Precision_Dict[a_bits])
-        next_out_tensor = next_layer.get_output(0)
-        # set the last layer's output tensor range
-        next_out_tensor.dynamic_range = (tracked_min_output, tracked_max_output)
-
-def build_engine(model_file, config=None, extra_layer_bits=32, strict_datatype=False, calib=None):
+    if 'output_bits' in config:
+        assert 'tracked_min_output' in config
+        assert 'tracked_max_output' in config
+        act_bits = config['output_bits']
+        tracked_min_output = config['tracked_min_output']
+        tracked_max_output = config['tracked_max_output']
+        layer.set_output_type(0, Precision_Dict[act_bits])
+        out_tensor = layer.get_output(0)
+        out_tensor.dynamic_range = (tracked_min_output, tracked_max_output)
+
+def propagate_from_low_bit_predecessor(layer, out2layer, default_precision=trt.float16):
+    """
+    Returns
+    -------
+    layer precision
+        current layer's precision
+    (min, max)
+        dynamic range of current layer's output tensor
     """
-    This function builds an engine from an onnx model with calibration process.
+    dynamic_range = None
+    tensor = layer.get_input(0)
+    if tensor is not None:
+        predecessor = out2layer[tensor.name]
+        if predecessor.get_output_type(0) == trt.int8:
```

Comment:
do we only support int8 right now?

Reply:
yes, only support int8 for now, added comments to make it clear.